src/wallet.rs

```diff
@@ -0,0 +1,962 @@
+
+        Ok((path.with_extension("key"), path.with_extension("cpk")))
+    }
+
+    /// Obtain the owned `Address` for a given address
+    pub fn claim_as_address(&self, addr: Address) -> Result<Address, Error> {
+        let my_addrs = self.addresses();
+        for a in my_addrs {
+            if a.psk == addr.psk {
+                return Ok(a);
+            }
+        }
+        Err(Error::AddressNotOwned)
+    }
+}
+
+/// A public address within the Dusk Network
+#[derive(Clone, Eq)]
+pub struct Address {
+    index: Option<u64>,
+    psk: PublicSpendKey,
+}
+
+impl Address {
+    pub(crate) fn new(index: u64, psk: PublicSpendKey) -> Self {
+        Self {
+            index: Some(index),
+            psk,
+        }
+    }
+
+    /// Returns true if the current user owns this address
+    pub fn is_owned(&self) -> bool {
+        self.index.is_some()
+    }
+
+    pub(crate) fn psk(&self) -> &PublicSpendKey {
+        &self.psk
+    }
+
+    pub(crate) fn index(&self) -> &Option<u64> {
+        &self.index
+    }
+
+    /// A trimmed version of the address to display as preview
+    pub fn preview(&self) -> String {
+        let addr = bs58::encode(self.psk.to_bytes()).into_string();
+        (&addr[..10]).to_string()
+    }
+}
+
+impl FromStr for Address {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let dec = bs58::decode(s).into_vec()?;
+        if dec.len() != SEED_SIZE {
+            return Err(Error::BadAddress);
+        }
+        let mut addr_bytes = [0u8; SEED_SIZE];
+        addr_bytes.copy_from_slice(&dec);
+        let addr = Address {
+            index: None,
+            psk: dusk_pki::PublicSpendKey::from_bytes(&addr_bytes)?,
+        };
+        Ok(addr)
+    }
+}
+
+impl TryFrom<String> for Address {
+    type Error = Error;
+
+    fn try_from(s: String) -> Result<Self, Self::Error> {
+        let dec = bs58::decode(s).into_vec()?;
+        if dec.len() != SEED_SIZE {
+            return Err(Error::BadAddress);
+        }
+        let mut addr_bytes = [0u8; SEED_SIZE];
+        addr_bytes.copy_from_slice(&dec);
+        let addr = Address {
+            index: None,
+            psk: dusk_pki::PublicSpendKey::from_bytes(&addr_bytes)?,
+        };
+        Ok(addr)
+    }
+}
+
+impl TryFrom<&[u8; 64]> for Address {
+    type Error = Error;
+
+    fn try_from(bytes: &[u8; 64]) -> Result<Self, Self::Error> {
+        let addr = Address {
+            index: None,
+            psk: dusk_pki::PublicSpendKey::from_bytes(bytes)?,
+        };
+        Ok(addr)
+    }
+}
+
+impl PartialEq for Address {
+    fn eq(&self, other: &Self) -> bool {
+        self.index == other.index && self.psk == other.psk
+    }
+}
+
+impl std::hash::Hash for Address {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.index.hash(state);
+        self.psk.to_bytes().hash(state);
+    }
+}
+
+impl Display for Address {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
```

Comment:
You should probably import `std::fmt`, it would shorten the syntax to `fmt::Formatter` and `fmt::Result`, that usually the idiomatic way.